python/mlir_structured/dialects/indexing.py

```diff
@@ -366,10 +370,268 @@ def __class_getitem__(
         static_sizes.append(ShapedType.get_dynamic_size())
     return RankedTensorType.get(static_sizes, dtype)
 
-  def __getitem__(self, dims: tuple) -> Scalar:
-    dims = list(dims)
-    for i, d in enumerate(dims):
+  def __getitem__(self, idx: tuple) -> Scalar:
+    # Early bail for most trivial corner cases (ellipse or one slice(None))
+    if idx == Ellipsis or idx == slice(None):
+      return self
+    idx = list((idx,) if isinstance(idx, int) else idx)
+    for i, d in enumerate(idx):
       if isinstance(d, int):
-        dims[i] = arith.ConstantOp.create_index(d).result
+        idx[i] = Scalar(arith.ConstantOp.create_index(d))
 
-    return Scalar(tensor.ExtractOp(self, dims))
+    if all(isinstance(d, Scalar) and d.is_constant()
+           for d in idx) and len(idx) == len(self.shape):
+      return Scalar(tensor.ExtractOp(self, idx))
+    else:
+      return build_gather(self, tuple(idx))
+
+
+#################
+# advanced indexing
+#################
+
+
+class _Indexer(NamedTuple):
+  """Data that describes/captures the relationship between an indexing tensor
+  (produced from simple or advanced indices) and scatter/gather dims.
+
+  Args:
+    indices: The final index tensor.
+    collapsed_dims: The set of dimensions i in operand that have
+      shape[i] == 1. Must be a tuple of integers in ascending order.
+      These determine the gather or scatter dims.
+    newaxis_dims: axes created by `np.newaxis` or None. These must be inserted
+      for gathers and eliminated for scatters.
+    unique_indices: Describes whether indices known to be non-overlapping.
+  """
+
+  indices: Tensor
+  collapsed_dims: Tuple[int, ...]
+  newaxis_dims: Tuple[int, ...]
+  unique_indices: Optional[bool] = None
+
+
+def _is_empty(ten: Tensor) -> bool:
+  """Check whether any of the dimensions of `ten` are degenerate (i.e., have
+  size 0).
+
+  Args:
+    ten: Tensor whose shape is under consideration.
+
+  Returns:
+    Whether any of the dimensions of `ten` are degenerate.
+  """
+  shape = ten.shape
+  if isinstance(shape,
+                (tuple, list)) and all(isinstance(s, int) for s in shape):
+    return any(s == 0 for s in shape)
+  raise NotImplementedError(shape)
+
+
+def concatenate(tens: Sequence[Tensor], dim):
+  """Concatenate a sequence Tensors along dimension `dim`.
+
+  Args:
+    tens: Sequence of tensors all having the same shape, except for along
+      dimension `dim`.
+    dim: Dimension to concatenate along.
+
+  Returns:
+    Tensor that wraps a value that's either the result of index.concatenate
+    or (if all tensors are constant and foldable) the result of arith.constant.
+  """
+  if len(tens) <= 1:
+    raise ValueError(f"Expected multiple {tens=}")
```

Comment:
Is this a necessary/useful restriction? The unix tool `cat` is also able to "concatenate" a single file...

Reply:
Turns out it's not: 

```python
concat_single_ten_first_dim = indexing.ConcatenateOp((ten,), 0).result
print(concat_single_ten_first_dim.owner)
```
```
%1 = indexing.concatenate(%0) {dim = 0} : (tensor<10x10xi32>) -> tensor<10x10xi32>
```